### PR TITLE
Add ability to query EC2 ENI instance limits via ec2:DescribeInstanceTypes

### PR DESCRIPTION
**Pull request by ungureanuvladvictor**
AWS added one new endpoint to the EC2 API that allows describing all the
available EC2 instance types in the region (https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstanceTypes.html).

This allows us to rely on this going forward instead of keeping a
hardcoded map of instance types to adapter limits.

This will require a new EC2 API persmission to be added to the default
IAM permission set: ec2:DescribeInstanceTypes.

Signed-off-by: Vlad Ungureanu <vladu@palantir.com>

```release-note
Add ability to query EC2 ENI instance limits via ec2:DescribeInstanceTypes
```

Main downside with this PR is that the `ec2:DescribeInstanceTypes` API call is available just in the v1 of the `aws-sdk-go` and not in the v2 that the `cilium-operator` uses.

@tgraf opinions on this since we talked back in the day about vendoring the v1 or v2 SDK (when initial ENI support was built). I'm also fine waiting for the v2 SDK to get support for the endpoint that we need and update afterwards.

Would like to get https://github.com/cilium/cilium/pull/9693 in first and then update the docs as well via this one.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9699)

### CODE REVIEW

**Comment by ungureanuvladvictor** [comment]
test-me-please

**Comment by ungureanuvladvictor** [comment]
test-me-please

**Comment by ungureanuvladvictor** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
Docs changes LGTM.

**Comment by joestringer** [comment]
test-docs-please

**Review by vadorovsky — APPROVED** [review]

**Comment by joestringer** [comment]
Looking for feedback from @tgraf before merge, since he contributed the ENI feature and he's called out in the feature description.

**Review by aanm — COMMENTED** [review]
@ungureanuvladvictor did you downgrade the aws library already?

**Comment by ungureanuvladvictor** [comment]
@aanm what do you mean by downgrading it? As I mentioned on slack the v2 sdk does not have support for this endpoint so that is why I'm vendoring v1 explicitly + also upgrading to latest.

File: go.mod [hunk 1]
```diff
@@ -141,7 +142,7 @@ replace (
 	github.com/armon/go-radix => github.com/armon/go-radix v0.0.0-20180808171621-7fddfc383310
 	github.com/asaskevich/govalidator => github.com/asaskevich/govalidator v0.0.0-20190424111038-f61b66f89f4a
 	github.com/auth0/go-jwt-middleware => github.com/auth0/go-jwt-middleware v0.0.0-20170425171159-5493cabe49f7
-	github.com/aws/aws-sdk-go => github.com/aws/aws-sdk-go v1.16.26
+	github.com/aws/aws-sdk-go => github.com/aws/aws-sdk-go v1.25.43
 	github.com/aws/aws-sdk-go-v2 => github.com/aws/aws-sdk-go-v2 v0.14.0
```
Comment (aanm) [inline go.mod:146]: @ungureanuvladvictor I thought you would be removing this dependency in favor of `github.com/aws/aws-sdk-go`

Comment (ungureanuvladvictor) [reply go.mod:146]: I was not planning on doing that. Happy to do it if you think that would be helpful overall. In their README they call out:
```
aws-sdk-go-v2 is the Developer Preview (aka beta) for the v2 AWS SDK for the Go programming language. This Developer Preview is provided to receive feedback from the language community on SDK changes prior to the final release. As such users should expect the SDK to release minor version releases that break backwards compatability
```

**Review by tgraf — COMMENTED** [review]
@ungureanuvladvictor Do we have an estimate what the timeframe of the API call hitting the v2 API? What is the impact on the binary size with this PR? That would be my only reservation and potential motivation to wait for this to hit the v2 API.

**Comment by ungureanuvladvictor** [comment]
@tgraf `cilium-operator` built from current HEAD (f0c985451cb064e6103a0b05349c4dfb4ef780b9) is 71M vs. my branch with the v1 SDK is 83M. I realize this is not a small increase but tbh don't know when AWS folks will decide to promote this API to their V2 sdk.

**Comment by tgraf** [comment]
@ungureanuvladvictor I thought some more about this and I'm OK with this. I think we should be splitting the operator into individual pieces so we can run a container that takes care of cloud provider-specific aspects so non-AWS users do not have to pull in the AWS SDK dependencies.

@aanm @joestringer Are you OK with this regarding the binary size implications?

**Comment by aanm** [comment]
> @ungureanuvladvictor I thought some more about this and I'm OK with this. I think we should be splitting the operator into individual pieces so we can run a container that takes care of cloud provider-specific aspects so non-AWS users do not have to pull in the AWS SDK dependencies.
> 
> @aanm @joestringer Are you OK with this regarding the binary size implications?

All good to me, the binary size decreased by the removal of `pkg/workloads`

**Pull request merged**